### PR TITLE
Add Listener Plugin API that tracks TaskInstance state changes

### PR DESCRIPTION
**Pull request by mobuchowski**
This PR adds new Plugin API - listeners. It enables plugin authors to write [`pluggy` hook implementation](https://pluggy.readthedocs.io/en/stable/) that will be called on certain formalized extension points. To differentiate between current Airflow extension points, like plugins, and current Airflow hooks, implementations of those hooks are called `listeners`.

The API is ment to be called across all dags, and all operators - in contrast to current `on_success_callback`, `pre_execute` and related family which are meant to provide callbacks for particular dag authors, or operator creators. 

`pluggy` mechanism enables us to execute multiple, or none, listeners that implement particular extension point, so that users can use multiple listeners seamlessly.

In this PR, three such extension points are added. 
When TaskInstance's state is changed to `RUNNING`, `on_task_instance_running` hook is called. On change to`SUCCESS` `on_task_instance_success` is called, similarly on `FAILED` `on_task_instance_failed` is called.

Actual notification mechanism is be implemented using [SQLAlchemy’s events mechanism](https://docs.sqlalchemy.org/en/13/orm/session_events.html#after-flush). This ensures that plugins will get every change of state, regardless of where in the codebase it happened, and not require manual annotation of TI state changes across the codebase.

To make sure that this change is not affecting performance, running this mechanism on scheduler is disabled by default. The SQLAlchemy event mechanism is also not affected by default - the event listener is only added if we have any plugin which actually provides any listener. 

Design doc:
https://docs.google.com/document/d/1L3xfdlWVUrdnFXng1Di4nMQYQtzMfhvvWDR9K4wXnDU/edit?usp=sharing

Background to this change: 
When discussing changing LineageBackend api for OpenLineage plugin, we came to consensus that the changes proposed there would be better suited to the new API https://github.com/apache/airflow/issues/17984 

related: https://github.com/apache/airflow/issues/17984

### CODE REVIEW

**Comment by mobuchowski** [comment]
FYI @bolkedebruin @potiuk @julienledem

**Comment by ashb** [comment]
I think this change is small enough and isolated enough not need an AIP

File: airflow/utils/orm_event_handlers.py [hunk 1]
```diff
@@ -86,3 +86,7 @@ def after_cursor_execute(conn, cursor, statement, parameters, context, executema
                 stack_info,
                 statement.replace("\n", " "),
             )
+
+    if conf.getboolean('core', 'execute_listeners_on_scheduler', fallback=False):
+        # On import, register sqlalchemy event handlers which call airflow.listeners
+        import airflow.listeners.events  # noqa
```
Comment (uranusjr) [inline airflow/utils/orm_event_handlers.py]: IMO this is generally bad practice. I’d much prefer the import to not have side effect, and an explicit init function call registers the handlers. This would make tests much easier to write as well since you can do the init in a module-scoped pytest fixture instead.

Comment (potiuk) [reply airflow/utils/orm_event_handlers.py]: I definitely agree with @uranusjr . We already have some of those in Airflow and side effects have bitten us more than once.

Comment (mobuchowski) [reply airflow/utils/orm_event_handlers.py]: Fixed. Now it explicitely calls `register_task_instance_state_events`.

File: airflow/listeners/events.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        Listens for session.flush() events that modify TaskInstance's state, and notify listeners that listen
+        for that event. Doing it this way enable us to be stateless in the SQLAlchemy event listener.
+        """
+        for state in flush_context.states:
+            if isinstance(state.object, TaskInstance) and session.is_modified(
+                state.object, include_collections=False
+            ):
+                added, unchanged, deleted = flush_context.get_attribute_history(state, 'state')
+
+                logger.debug(f"session flush listener: added {added} unchanged {unchanged} deleted {deleted}")
```
Comment (kaxil) [inline airflow/listeners/events.py]:
```suggestion
                logger.debug("Session flush listener: added %s unchanged %s deleted %s", added, unchanged, deleted)
```

File: airflow/jobs/local_task_job.py [hunk 3]
```diff
@@ -291,3 +293,19 @@ def _update_dagrun_state_for_paused_dag(self, session=None):
...
+        Check if we have any registered listeners, then register sqlalchemy hooks for
+        TI state change if we do.
+        """
+        from airflow.plugins_manager import integrate_listener_plugins
+
+        integrate_listener_plugins()
+        from airflow.listeners.listener import get_listener_manager
+
+        if get_listener_manager().has_listeners():
+            from airflow.listeners.events import register_task_instance_state_events
```
Comment (kaxil) [inline airflow/jobs/local_task_job.py]: Is there a reason why imports are in such order, if yes, can we add comments explaining that please

File: airflow/listeners/listener.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+class Listener:
+    """Class used as a namespace for listener hook implementation namespace"""
+
+
+_listener_manager = None
+
+
+class ListenerManager:
+    """Class that manager registration of listeners and provides hook property for calling them"""
```
Comment (kaxil) [inline airflow/listeners/listener.py]:
```suggestion
    """Class that manages registration of listeners and provides hook property for calling them"""
```

File: airflow/listeners/listener.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    def __init__(self):
+        self.pm = pluggy.PluginManager("airflow")
+        self.pm.add_hookspecs(spec)
+        self.listener_names = set()
+
+    def has_listeners(self) -> bool:
+        return len(self.pm.get_plugins()) > 0
+
+    @property
+    def hook(self):
```
Comment (kaxil) [inline airflow/listeners/listener.py]: Can we add TypeHints for this property

File: airflow/listeners/listener.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+_listener_manager = None
+
+
+class ListenerManager:
+    """Class that manager registration of listeners and provides hook property for calling them"""
+
+    def __init__(self):
+        self.pm = pluggy.PluginManager("airflow")
+        self.pm.add_hookspecs(spec)
+        self.listener_names = set()
```
Comment (kaxil) [inline airflow/listeners/listener.py]:
```suggestion
        self.listener_names: Set[str] = set()
```

File: airflow/utils/orm_event_handlers.py [hunk 7]
```diff
@@ -86,3 +86,9 @@ def after_cursor_execute(conn, cursor, statement, parameters, context, executema
                 stack_info,
                 statement.replace("\n", " "),
             )
+
+    if conf.getboolean('core', 'execute_listeners_on_scheduler', fallback=False):
```
Comment (kaxil) [inline airflow/utils/orm_event_handlers.py]:
```suggestion
    if settings.EXECUTE_LISTENERS_ON_SCHEDULER:
```

File: airflow/settings.py [hunk 8]
```diff
@@ -543,6 +543,9 @@ def initialize():
 # loaded from module.
 LAZY_LOAD_PROVIDERS = conf.getboolean('core', 'lazy_discover_providers', fallback=True)
 
+# By default running listeners on scheduler is disabled. Set it to True if you want to execute them.
+EXECUTE_LISTENERS_ON_SCHEDULER = conf.getboolean('core', 'execute_listeners_on_scheduler', fallback=False)
```
Comment (kaxil) [inline airflow/settings.py]: It needs to be added to https://github.com/apache/airflow/blob/main/airflow/config_templates/default_airflow.cfg and https://github.com/apache/airflow/blob/main/airflow/config_templates/config.yml too as we have got automated docs building from there - https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html

File: tests/listeners/test_listeners.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import os
+
+import pluggy
+import pytest as pytest
+
+os.environ["AIRFLOW__CORE__EXECUTE_LISTENERS_ON_SCHEDULER"] = "True"
```
Comment (kaxil) [inline tests/listeners/test_listeners.py]: This will cause issues if for some reason this file is imported by any other tests, unlikely but still. 

Can we instead use a pytest module level fixture and patch `os.environ`

File: airflow/settings.py (hunk 8 above)
Comment (mobuchowski) [reply airflow/settings.py]: Done. Also, added basic docs.

File: airflow/utils/orm_event_handlers.py (hunk 7 above)
Comment (mobuchowski) [reply airflow/utils/orm_event_handlers.py]: Done.

File: airflow/listeners/listener.py (hunk 6 above)
Comment (mobuchowski) [reply airflow/listeners/listener.py]: Done.

File: airflow/listeners/listener.py (hunk 5 above)
Comment (mobuchowski) [reply airflow/listeners/listener.py]: Done.

File: airflow/listeners/listener.py (hunk 4 above)
Comment (mobuchowski) [reply airflow/listeners/listener.py]: Done.

File: airflow/listeners/events.py (hunk 2 above)
Comment (mobuchowski) [reply airflow/listeners/events.py]: Done.

File: airflow/jobs/local_task_job.py (hunk 3 above)
Comment (mobuchowski) [reply airflow/jobs/local_task_job.py]: I'm only importing `register_task_instance_state_events` now - locally, to prevent circular dependencies.

File: tests/listeners/test_listeners.py (hunk 9 above)
Comment (mobuchowski) [reply tests/listeners/test_listeners.py]: Replaced with

```
@pytest.fixture(scope="module", autouse=True)
def register_events():
    register_task_instance_state_events()
```

File: airflow/__init__.py [hunk 10]
```diff
@@ -80,6 +80,11 @@ def __getattr__(name):
     manager.initialize_providers_hooks()
     manager.initialize_providers_extra_links()
 
+if settings.EXECUTE_LISTENERS_ON_SCHEDULER:
+    from airflow.plugins_manager import integrate_listener_plugins
+
+    integrate_listener_plugins()
```
Comment (ashb) [inline airflow/__init__.py]: Given the name of this setting I wonder if this code should be somewhere in SchedulerJob instead?

File: airflow/jobs/local_task_job.py [hunk 11]
```diff
@@ -291,3 +296,13 @@ def _update_dagrun_state_for_paused_dag(self, session=None):
...
                 dag_run.update_state(session=session, execute_callbacks=True)
+
+    @staticmethod
+    def _enable_task_listeners():
+        """
+        Check if we have any registered listeners, then register sqlalchemy hooks for
+        TI state change if we do.
+        """
+        integrate_listener_plugins()
+        if get_listener_manager().has_listeners():
```
Comment (ashb) [inline airflow/jobs/local_task_job.py]: `integrate_listener_plugins` should be called automatically from inside `get_listener_manager` I think.

File: airflow/jobs/local_task_job.py [hunk 12]
```diff
@@ -71,6 +74,8 @@ def __init__(
         # terminate multiple times
         self.terminating = False
 
+        self._enable_task_listeners()
```
Comment (ashb) [inline airflow/jobs/local_task_job.py]: I think should should be moved in to `_execute` instead of the constructor -- it (currently) doesn't make much difference, to anything other than tests, but currently the constructor otherwise doesn't do anything than create an object, so lets keep it that way.

File: airflow/listeners/events.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+from airflow.listeners.listener import get_listener_manager
+from airflow.models import TaskInstance
+from airflow.utils.state import State
+
+_is_listening = False
+
+
+def register_task_instance_state_events():
+    logger = logging.getLogger()
```
Comment (ashb) [inline airflow/listeners/events.py]:
```suggestion
    logger = logging.getLogger(__name__)
```

File: airflow/listeners/events.py [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+            ):
+                added, unchanged, deleted = flush_context.get_attribute_history(state, 'state')
+
+                logger.warning(
+                    "session flush listener: added %s unchanged %s deleted %s - %s",
+                    added,
+                    unchanged,
+                    deleted,
+                    state.object,
+                )
```
Comment (ashb) [inline airflow/listeners/events.py]: Move this to debug level now?

File: airflow/listeners/listener.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+class ListenerManager:
+    """Class that manages registration of listeners and provides hook property for calling them"""
+
+    def __init__(self):
+        self.pm = pluggy.PluginManager("airflow")
+        self.pm.add_hookspecs(spec)
+        self.listener_names: Set[str] = set()
+
+    def has_listeners(self) -> bool:
```
Comment (ashb) [inline airflow/listeners/listener.py:45]: Maybe make this a property?

```suggestion
    @property
    def has_listeners(self) -> bool:
```

File: airflow/listeners/listener.py [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+import inspect
+import logging
+from typing import TYPE_CHECKING, Set
+
+import pluggy
+
+if TYPE_CHECKING:
+    from pluggy._hooks import _HookRelay
+
+from airflow.listeners import spec
```
Comment (ashb) [inline airflow/listeners/listener.py]: Delay this import until ListenerManager init?

File: airflow/listeners/listener.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+            return
+
+        listener_type = type(listener)
+        if not (
+            inspect.isclass(listener_type)
+            and issubclass(listener_type, Listener)
+            and (listener_type is not Listener)
+        ):
+            log.warning("Can't register listener: %s - is not a Listener subclass", listener_type)
+            return
```
Comment (ashb) [inline airflow/listeners/listener.py]: We shouldn't need to do any of this -- one of the main features of Pluggy is that it handles signature validation for us!

https://pluggy.readthedocs.io/en/latest/#enforcing

File: airflow/listeners/spec.py [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+    from airflow.models import TaskInstance
+    from airflow.utils.state import State
+
+hookspec = HookspecMarker("airflow")
+
+
+@hookspec
+def on_task_instance_running(
+    previous_state: "State", task_instance: "TaskInstance", session: Optional["Session"]
```
Comment (ashb) [inline airflow/listeners/spec.py]:
```suggestion
    previous_state: "TaskInstanceState", task_instance: "TaskInstance", session: Optional["Session"]
```

in a few places.

File: docs/apache-airflow/plugins.rst [hunk 19]
```diff
@@ -268,6 +285,7 @@ definitions in Airflow.
         operator_extra_links = [
             S3LogLink(),
         ]
+        listeners = [PrintingRunningListener]
```
Comment (ashb) [inline docs/apache-airflow/plugins.rst]: I'm not sure about the need for the class here. Is there a reason we don't follow the "pluggy" way here and do:

```suggestion
        listeners = [on_task_instance_running]
```

File: tests/jobs/test_scheduler_job.py [hunk 20]
```diff
@@ -3434,6 +3435,12 @@ def test_timeout_triggers(self, dag_maker):
         assert ti1.next_method == "__fail__"
         assert ti2.state == State.DEFERRED
 
+    @pytest.fixture(autouse=True, scope='function')
+    def clean_listener(self):
+        get_listener_manager().clear()
+        yield
+        get_listener_manager().clear()
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py]: Why is this needed in SchedulerJob tests?

File: docs/apache-airflow/plugins.rst (hunk 19 above)
Comment (ashb) [reply docs/apache-airflow/plugins.rst]: The disadvantage of using a class is it might imply that you can store state in the instance between plugin invocations (which isn't the case, not in general terms anyway) where as a function might make that clearer.

File: docs/apache-airflow/plugins.rst [hunk 21]
```diff
@@ -167,9 +172,14 @@ definitions in Airflow.
 
     # Importing base classes that we need to derive
     from airflow.hooks.base import BaseHook
+    from airflow.listeners.listener import Listener
     from airflow.models.baseoperator import BaseOperatorLink
     from airflow.providers.amazon.aws.transfers.gcs_to_s3 import GCSToS3Operator
 
+    from pluggy import HookimplMarker
+
+    hookimpl = HookimplMarker("airflow")
```
Comment (ashb) [inline docs/apache-airflow/plugins.rst]: Looking at the pluggy docs it appears the normal pattern here is to import the hookimpl from a module rather than to create it here: https://pluggy.readthedocs.io/en/latest/#the-plugin

File: tests/listeners/test_listeners.py [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+def test_listener_captures_failed_taskinstances(create_task_instance_of_operator, session=None):
+    lm = get_listener_manager()
+    listener = FullListener()
+    lm.add_listener(listener)
+
+    with pytest.raises(AirflowException):
+        ti = create_task_instance_of_operator(
+            BashOperator, dag_id=DAG_ID, execution_date=EXECUTION_DATE, task_id=TASK_ID, bash_command="exit 1"
+        )
+        ti.run()
```
Comment (ashb) [inline tests/listeners/test_listeners.py]: Hmmm, if we can avoid calling `ti.run()` here in this test that would be good. (It does _a lot_ and is relatively heavy weight approach, so if we can do something lighter here that would be preferable)

File: tests/plugins/test_plugins_manager.py [hunk 23]
```diff
@@ -350,6 +357,21 @@ class MacroPlugin(AirflowPlugin):
...
+            plugins_manager.load_plugins_from_plugin_directory()
+            plugins_manager.integrate_listener_plugins()
+
+            assert get_listener_manager().has_listeners()
+            assert get_listener_manager().pm.get_plugins().pop().__class__.__name__ == "PluginListener"
+
+    @pytest.fixture(autouse=True)
+    def clear_listeners(self):
+        yield
+        get_listener_manager().clear()
```
Comment (ashb) [inline tests/plugins/test_plugins_manager.py]:
```suggestion
```

Already have this as a module level fixture.

**Review by ashb — CHANGES_REQUESTED** [review]
Looking good overall, a few changes/questions to address.

File: docs/apache-airflow/plugins.rst (hunk 19 above)
Comment (mobuchowski) [reply docs/apache-airflow/plugins.rst]: Generally, in pluggy, we need to register `plugin` that's a namespace that consists of possibly multiple hookimpls, not single functions. 
By passing just a list of functions, we'd later need to construct "synthetic" namespace for it to register.

I think requiring a class, or, possibly a python module, is a good idea - this is what pluggy docs present:

- class https://pluggy.readthedocs.io/en/stable/index.html#a-toy-example
- module https://pluggy.readthedocs.io/en/stable/index.html#the-plugin

Now, it does not need to look like 
```python
        listeners = [PrintingRunningListener]
```

We could require plugin authors to directly implement hookimpls in plugins, for example

```python
class SomeListenerPlugin(AirflowPlugin):
    name = 'some-plugin'
    @hookimpl
    def on_task_instance_running(...):
        print("running")
``` 

or, we could require it to add single pluggy plugin - as class or module - if for some reason we don't want them to implement multiple listener functions in single plugin.

I agree that we should discourage trying to keep state in plugin - my tests do not help, as I do that there :)
Changing the tests, and mentioning that directly in docs might be a good idea.

Note, this is separate discussion from https://github.com/apache/airflow/pull/20443#discussion_r778047132

Comment (ashb) [reply docs/apache-airflow/plugins.rst]: Oh yes, you are right. Thinking... 🤔

File: airflow/__init__.py (hunk 10 above)
Comment (mobuchowski) [reply airflow/__init__.py]: Actually, I've removed it for now. It will be easy to add it later.

File: airflow/jobs/local_task_job.py (hunk 12 above)
Comment (mobuchowski) [reply airflow/jobs/local_task_job.py]: Done.

File: airflow/listeners/spec.py (hunk 18 above)
Comment (mobuchowski) [reply airflow/listeners/spec.py]: Done.

File: airflow/jobs/local_task_job.py (hunk 11 above)
Comment (mobuchowski) [reply airflow/jobs/local_task_job.py]: Done. Changed `integrate_listener_plugins` to require `ListenerManager` parameter since they would have circular calls otherwise.

File: airflow/listeners/events.py (hunk 13 above)
Comment (mobuchowski) [reply airflow/listeners/events.py]: Done.

File: airflow/listeners/events.py (hunk 14 above)
Comment (mobuchowski) [reply airflow/listeners/events.py]: Done.

File: airflow/listeners/listener.py (hunk 15 above)
Comment (mobuchowski) [reply airflow/listeners/listener.py:45]: Done.

File: airflow/listeners/listener.py (hunk 16 above)
Comment (mobuchowski) [reply airflow/listeners/listener.py]: Done, importing this on `__init__` now.

File: airflow/listeners/listener.py (hunk 17 above)
Comment (mobuchowski) [reply airflow/listeners/listener.py]: Done. Right now `add_listener` accepts anything that `pluggy` would.

File: tests/jobs/test_scheduler_job.py (hunk 20 above)
Comment (mobuchowski) [reply tests/jobs/test_scheduler_job.py]: Removed. Artifact from debugging...

File: tests/listeners/test_listeners.py (hunk 22 above)
Comment (mobuchowski) [reply tests/listeners/test_listeners.py]: I can do `_run_raw_task`.

File: tests/plugins/test_plugins_manager.py (hunk 23 above)
Comment (mobuchowski) [reply tests/plugins/test_plugins_manager.py]: Removed.

File: docs/apache-airflow/plugins.rst (hunk 21 above)
Comment (mobuchowski) [reply docs/apache-airflow/plugins.rst]: Done.

File: docs/apache-airflow/plugins.rst (hunk 19 above)
Comment (ashb) [reply docs/apache-airflow/plugins.rst]: So looking at pytest itself (which is where pluggy came out of) the plugins are def just modules, not classes https://docs.pytest.org/en/6.2.x/writing_plugins.html

I think we should follow pytest's lead here unless we have a very good reason to diverge.

Comment (mobuchowski) [reply docs/apache-airflow/plugins.rst]: @ashb added check for modules in `add_listener.py` and refactored tests from classes to modules.

File: docs/apache-airflow/listeners.rst [hunk 24]
```diff
@@ -0,0 +1,39 @@
...
+
+Airflow gives you an option to be notified of events happening in Airflow
+by writing listeners. Listeners are powered by `pluggy <https://pluggy.readthedocs.io/en/stable/>`__
+
+Listener API is meant to be called across all dags, and all operators - in contrast to methods like
+``on_success_callback``, ``pre_execute`` and related family which are meant to provide callbacks
+for particular dag authors, or operator creators. There is no possibility to listen on events generated
+by particular dag.
+
+To include listener in your Airflow installation, include it as a part of an :doc:`Airflow Plugin </plugins>`
```
Comment (ashb) [inline docs/apache-airflow/listeners.rst:29]:
```suggestion
To include listener in your Airflow installation, include it as a part of an :doc:`Airflow Plugin </plugins>`

|experimental|
```

This is a substitution that just links to experimental page, gives us some wiggle room if we need to change anything here in next release without worrying about back-compat.

File: tests/listeners/test_empty_listener.py [hunk 25]
```diff
@@ -0,0 +1,24 @@
+#
```
Comment (ashb) [inline tests/listeners/test_empty_listener.py]: Let's change the name of these files that don't contain tests to just `empty_listener` etc -- the `test_` prefix is confusing here as they don't contain any tests themselves, but are used by them.

File: tests/listeners/test_listeners.py [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+from tests.listeners import test_full_listener, test_partial_listener, test_throwing_listener
+
+DAG_ID = "test_listener_dag"
+TASK_ID = "test_listener_task"
+EXECUTION_DATE = timezone.utcnow()
+
+
+@pytest.fixture(scope="module", autouse=True)
+def register_events():
+    register_task_instance_state_events()
```
Comment (ashb) [inline tests/listeners/test_listeners.py]: Do we need to remove this setting/config when the test finishes?

File: airflow/plugins_manager.py [hunk 27]
```diff
@@ -458,6 +463,20 @@ def integrate_macros_plugins() -> None:
             setattr(macros, plugin.name, macros_module)
 
 
+def integrate_listener_plugins(listener_manager: "ListenerManager") -> None:
+    global plugins
+    global macros_modules
```
Comment (ashb) [inline airflow/plugins_manager.py]:
```suggestion
```

Not used?

File: tests/listeners/test_listeners.py [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+    test_full_listener.state = []
+
+
+@provide_session
+def test_listener_gets_calls(create_task_instance, session=None):
+    lm = get_listener_manager()
+    lm.add_listener(test_full_listener)
+
+    ti = create_task_instance(session=session, state=State.QUEUED)
+    ti.run()
```
Comment (ashb) [inline tests/listeners/test_listeners.py:60]: Is this one using `ti.run()` and not `ti._run_raw_task()` for a reason? If so please add a short comment

File: tests/listeners/test_listeners.py [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+@provide_session
+def test_listener_captures_failed_taskinstances(create_task_instance_of_operator, session=None):
+    lm = get_listener_manager()
+    lm.add_listener(test_full_listener)
+
+    with pytest.raises(AirflowException):
+        ti = create_task_instance_of_operator(
+            BashOperator, dag_id=DAG_ID, execution_date=EXECUTION_DATE, task_id=TASK_ID, bash_command="exit 1"
+        )
+        ti._run_raw_task()
```
Comment (ashb) [inline tests/listeners/test_listeners.py]: Please scope the `raises` tighter (in a few places)

```suggestion

    ti = create_task_instance_of_operator(
        BashOperator, dag_id=DAG_ID, execution_date=EXECUTION_DATE, task_id=TASK_ID, bash_command="exit 1"
    )
    with pytest.raises(AirflowException):
            ti._run_raw_task()
```

File: tests/listeners/test_listeners.py [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def test_non_module_listener_is_not_registered():
+    class NotAListener:
+        @hookimpl
+        def on_task_instance_running(self, previous_state, task_instance, session):
+            pass
+
+    lm = get_listener_manager()
+    lm.add_listener(NotAListener())
```
Comment (ashb) [inline tests/listeners/test_listeners.py]: Shouldn't this throw a TypeError instead?

**Review by ashb — CHANGES_REQUESTED** [review]
Impl looks good to me now (and suprisingly small, well done!) just a few changes to the tests please.

File: tests/listeners/test_listeners.py (hunk 29 above)
Comment (mobuchowski) [reply tests/listeners/test_listeners.py]: Fixed.

File: tests/listeners/test_listeners.py (hunk 28 above)
Comment (mobuchowski) [reply tests/listeners/test_listeners.py:60]: Added comment:

```
    # Using ti.run() instead of ti._run_raw_task() to capture state change to RUNNING
    # that only happens on `check_and_change_state_before_execution()` that is called before
    # `run()` calls `_run_raw_task()`
```

File: airflow/plugins_manager.py (hunk 27 above)
Comment (mobuchowski) [reply airflow/plugins_manager.py]: Removed.

File: docs/apache-airflow/listeners.rst (hunk 24 above)
Comment (mobuchowski) [reply docs/apache-airflow/listeners.rst:29]: Done.

File: tests/listeners/test_empty_listener.py (hunk 25 above)
Comment (mobuchowski) [reply tests/listeners/test_empty_listener.py]: Done.

File: tests/listeners/test_listeners.py (hunk 26 above)
Comment (mobuchowski) [reply tests/listeners/test_listeners.py]: Added `unregister_task_instance_state_events` function - please take a look at `airflow.listeners.events` now.

File: tests/listeners/test_listeners.py (hunk 30 above)
Comment (mobuchowski) [reply tests/listeners/test_listeners.py]: Yes. Fixed.

**Review by ashb — APPROVED** [review]
